### PR TITLE
CXF-8089: Build Comma Separated Values in url from Array/List Query Param

### PR DESCRIPTION
**Pull request by reta**
### Description
The typical style the collection-like query parameters are encoded assumes the repetition of the parameter in question multiple times, for example:

    http://localhost/books?ids=1&ids=2&ids=3&ids=4

However, as part of https://issues.apache.org/jira/browse/CXF-6941, the CXF server-side includes support for collection query parameters encoded as comma-separated strings, for example:

    http://localhost/books?ids=1,2,3,4

This is quite useful in many cases however the CXF's client side does not anyhow support this encoding and always uses the repetition (`ids=1&ids=2&ids=3&ids=4`). This PRs complements the client-side feature set.

@coheigea @deki @andymc12 if you guys could take a look please, would appreciate the opinion, there are a few corner cases to cover but the idea in general seems to be working fine, thanks guys.

### CODE REVIEW

**Comment by deki** [comment]
Hi Andriy, nice improvement but I wonder what are the specs saying about this? Will it work only if client and server side are using CXF?

**Comment by reta** [comment]
Hi @deki,

Thanks a lot for the review. Yes, this is 100% CXF-specific. This is what spec (https://tools.ietf.org/html/rfc6570) says for multi-value query string:

  >  An explode modifier applied to a list variable causes the expansion
   to iterate over the list's member values.  For path and query
   parameter expansions, each member value is paired with the variable's
   name as a (varname, value) pair.  This allows path and query
   parameters to be repeated for multiple values, as in

     Given the variable assignments

       year  := ("1965", "2000", "2012")
       dom   := ("example", "com")

     Example Template     Expansion

       find{?year*}       find?year=1965&year=2000&year=2012
       www{.dom*}         www.example.com

Basically, CXF allows an alternative but adhere to spec by default. 
Thanks!

File: rt/rs/client/src/main/java/org/apache/cxf/jaxrs/client/WebClient.java [hunk 1]
```diff
@@ -90,11 +90,19 @@
     private static final String WEB_CLIENT_OPERATION_REPORTING = "enable.webclient.operation.reporting";
     private BodyWriter bodyWriter = new BodyWriter();
     protected WebClient(String baseAddress) {
-        this(convertStringToURI(baseAddress));
+        this(convertStringToURI(baseAddress), Collections.emptyMap());
```
Comment (andymc12) [inline rt/rs/client/src/main/java/org/apache/cxf/jaxrs/client/WebClient.java]: Same comment as before wrt Collections.emptyMap() - it would probably perform better if it were a constant. Same for lines 101 and 121.

File: rt/rs/client/src/main/java/org/apache/cxf/jaxrs/client/ClientProxyImpl.java [hunk 2]
```diff
@@ -119,7 +119,17 @@ public ClientProxyImpl(URI baseURI,
                            boolean isRoot,
                            boolean inheritHeaders,
                            Object... varValues) {
-        this(new LocalClientState(baseURI), loader, cri, isRoot, inheritHeaders, varValues);
+        this(baseURI, loader, cri, isRoot, inheritHeaders, Collections.emptyMap(), varValues);
```
[the remark this answers is not in the document]
Comment (reta) [reply rt/rs/client/src/main/java/org/apache/cxf/jaxrs/client/ClientProxyImpl.java]: Thanks a lot for review, @andymc12! The `Collections.emptyMap()` is actually backed by the constant, basically the only purpose of the method is to cast the generic parameters (this is implementation from `Collections`):
```
    public static final <K,V> Map<K,V> emptyMap() {
        return (Map<K,V>) EMPTY_MAP;
    }
```

Using `null` is also possible, but is a bit risky since `ClientProxyImpl` has a number of overloaded constructors.

Sounds good? Thanks a lot!

[the remark this answers is not in the document]
Comment (andymc12) [reply rt/rs/client/src/main/java/org/apache/cxf/jaxrs/client/ClientProxyImpl.java]: That makes sense, thanks.
